### ISSUE TITLE
Stop writing NULLs to EAV tables

### ISSUE THREAD

**Issue by loekvangool** [OpenMage/magento-lts#2997]
More of a discussion than a bug.

Magento/OpenMage writes `NULL` values in EAV tables (like `catalog_product_entity_*`) like it's going out of style. This creates significant additional bytes in the database without an apparent benefit.

The savings of deleting `NULL` in both data and indices  in our production database is easily 30%. Too big to ignore.

- Maybe the `catalog_product_entity_*.value` should be `NOT NULL`. Are there cases where a `NULL` be relevant vs. no record at all?
- Existing `NULL` values can be deleted without apparent loss of functionality

**Comment by loekvangool** [comment]
To see the changes for your database, you can run this. It will not affect the functionality to remove all `NULL` values (YMMV), and it will increase `AUTO INCREMENT`, but running it once is not a problem:

```sql
# Take size of EAV tables
SELECT TABLE_NAME AS `Table`, ROUND((DATA_LENGTH + INDEX_LENGTH) / 1024 / 1024) AS `Size (MB)`
FROM information_schema.TABLES
WHERE TABLE_SCHEMA = 'prokoeling' AND (`TABLE_NAME` LIKE 'catalog_product_entity_%' OR `TABLE_NAME` LIKE 'catalog_product_index_eav%')
ORDER BY (DATA_LENGTH + INDEX_LENGTH) DESC;

# Remove NULL values and rebuild table
DELETE FROM catalog_product_entity_int WHERE `value` IS NULL;
OPTIMIZE TABLE `catalog_product_entity_int`;

# Take size of EAV tables again
SELECT TABLE_NAME AS `Table`, ROUND((DATA_LENGTH + INDEX_LENGTH) / 1024 / 1024) AS `Size (MB)`
FROM information_schema.TABLES
WHERE TABLE_SCHEMA = 'prokoeling' AND (`TABLE_NAME` LIKE 'catalog_product_entity_%' OR `TABLE_NAME` LIKE 'catalog_product_index_eav%')
ORDER BY (DATA_LENGTH + INDEX_LENGTH) DESC;
```

**Comment by justinbeaty** [comment]
I ran on my local dev DB:

```mysql
mysql> DELETE FROM catalog_product_entity_int WHERE `value` IS NULL;
Query OK, 0 rows affected (0.11 sec)

mysql> DELETE FROM catalog_product_entity_text WHERE `value` IS NULL;
Query OK, 1223 rows affected (0.91 sec)

# I also optimized the two tables

mysql> SELECT TABLE_NAME AS `Table`, ROUND((DATA_LENGTH + INDEX_LENGTH) / 1024 / 1024) AS `Size (MB)` FROM information_schema.TABLES WHERE TABLE_SCHEMA = '__DATABASE__' AND (`TABLE_NAME` LIKE 'catalog_product_entity_%' OR `TABLE_NAME` LIKE 'catalog_product_index_eav%') ORDER BY (DATA_LENGTH + INDEX_LENGTH) DESC;
+--------------------------------------------+-----------+
| Table                                      | Size (MB) |
+--------------------------------------------+-----------+
| catalog_product_entity_text                |        42 |
| catalog_product_entity_int                 |        11 |
| catalog_product_index_eav                  |         9 |
| catalog_product_index_eav_idx              |         9 |
| catalog_product_entity_varchar             |         4 |
| catalog_product_entity_decimal             |         1 |
| catalog_product_entity_datetime            |         0 |
| catalog_product_entity_gallery             |         0 |
| catalog_product_entity_group_price         |         0 |
| catalog_product_entity_tier_price          |         0 |
| catalog_product_index_eav_decimal          |         0 |
| catalog_product_index_eav_decimal_idx      |         0 |
| catalog_product_entity_media_gallery       |         0 |
| catalog_product_entity_media_gallery_value |         0 |
| catalog_product_index_eav_decimal_tmp      |         0 |
| catalog_product_index_eav_tmp              |         0 |
+--------------------------------------------+-----------+
16 rows in set (0.00 sec)
```

There was no notable change, *but* I import all of my product data with magmi, so perhaps that avoids the null values as opposed to saving product info from the backend.

**Comment by justinbeaty** [comment]
One more datapoint, after dropping the null values, I went to the backend and saved a product (without making any changes). now:

```
mysql> select count(*) FROM catalog_product_entity_text WHERE `value` IS NULL;
+----------+
| count(*) |
+----------+
|       11 |
+----------+
1 row in set (0.10 sec)
```

So, Magento is just recreating the records for any empty text value.

I would presume fixing this to not write those records wouldn't have any negative affect, unless there's something going on with store values vs global values.

**Comment by justinbeaty** [comment]
One more thought:

If we did make it so magento-lts skips writing null for empty attributes, we would have to implement a way to delete records on from those tables on save. Otherwise you couldn't effectively unset a value. This may be obvious, but is a gotcha that need to be thought about, and is maybe why Magento works the way it does out of laziness.

**Comment by loekvangool** [comment]
> One more datapoint, after dropping the null values, I went to the backend and saved a product (without making any changes). now:
> 
> ```
> mysql> select count(*) FROM catalog_product_entity_text WHERE `value` IS NULL;
> +----------+
> | count(*) |
> +----------+
> |       11 |
> +----------+
> 1 row in set (0.10 sec)
> ```
> 
> So, Magento is just recreating the records for any empty text value.
> 
> I would presume fixing this to not write those records wouldn't have any negative affect, unless there's something going on with store values vs global values.
Yup that's what I have noticed as well. Your usual method of importing with magmi must prevent it somehow.

**Comment by justinbeaty** [comment]
It actually looks like mage has the ability to delete values. I'm not sure if this is the right approach at all, as I haven't tested it, but something like this could be the solution: https://github.com/justinbeaty/magento-lts/commit/59dbe35007bd49960a9e3fdf7ed62c85520026b4

**Comment by loekvangool** [comment]
Great, looking to hear from others on that code change.

As a datapoint, I removed ~300k `NULL` values from just `catalog_product_entity_int` which had ~1M rows before the removal.

~107 MB became ~75 MB with about 50/50 bytes for data and indices.

Of course the whole database should live in RAM, so expensive bytes, they are.

**Comment by justinbeaty** [comment]
> Great, looking to hear from others on that code change.

Just to reiterate, I haven't tested that code. I was mainly just curious because I started digging to see how magmi handled this, and then looked a bit into the Magento code. There might be good reasons for why Magento does it this way.
